client/app/bundles/system/admin/admin/components/tables/UsersTable.tsx

```diff
@@ -0,0 +1,371 @@
+  },
+  updateFailure: {
+    id: 'system.admin.user.update.failure',
+    defaultMessage: 'Failed to update user.',
+  },
+});
+
+const styles = {
+  list: {
+    paddingLeft: 0,
+    marginBottom: 0,
+  },
+  listItem: {
+    listStyle: 'none',
+  },
+};
+
+const UsersTable: FC<Props> = (props) => {
+  const { renderRowActionComponent, intl } = props;
+  const [isLoading, setIsLoading] = useState(false);
+  const users = useSelector((state: AppState) => getAllUserMiniEntities(state));
+  const counts = useSelector((state: AppState) => getAdminCounts(state));
+  const role = getUrlParameter('role');
+  const active = getUrlParameter('active');
+  const dispatch = useDispatch<AppDispatch>();
+
+  const [tableState, setTableState] = useState<TableState<UserMiniEntity>>({
+    count: counts.searchCount,
+    page: 0,
+    searchText: '',
+  });
+
+  useEffect((): void => {
+    setTableState({
+      ...tableState,
+      count: counts.searchCount,
+    });
+  }, [counts]);
+
+  const handleNameUpdate = (rowData, newName: string): Promise<void> => {
+    const user = rebuildObjectFromRow(
+      columns, // eslint-disable-line @typescript-eslint/no-use-before-define
+      rowData,
+    ) as UserMiniEntity;
+    const newUser = {
+      ...user,
+      name: newName,
+    };
+    return dispatch(updateUser(user.id, newUser))
+      .then(() => {
+        toast.success(
+          intl.formatMessage(translations.renameSuccess, {
+            oldName: user.name,
+            newName,
+          }),
+        );
+      })
+      .catch((error) => {
+        toast.error(intl.formatMessage(translations.updateFailure));
+        throw error;
+      });
+  };
+
+  const handleRoleUpdate = (
+    rowData,
+    newRole: string,
+    updateValue,
+  ): Promise<void> => {
+    const user = rebuildObjectFromRow(
+      columns, // eslint-disable-line @typescript-eslint/no-use-before-define
+      rowData,
+    ) as UserMiniEntity;
+    const newUser = {
+      ...user,
+      role: newRole as UserRole,
+    };
+    return dispatch(updateUser(user.id, newUser))
+      .then(() => {
+        updateValue(newRole);
+        toast.success(
+          intl.formatMessage(translations.changeRoleSuccess, {
+            name: user.name,
+            role: sharedConstants.USER_ROLES[newRole],
+          }),
+        );
+      })
+      .catch((error) => {
+        toast.error(intl.formatMessage(translations.updateFailure));
+        throw error;
+      });
+  };
+
+  const changePage = (page): void => {
+    setIsLoading(true);
+    setTableState({
+      ...tableState,
+      page,
+    });
+    dispatch(indexUsers({ page, role, active })).then(() => {
+      setIsLoading(false);
+    });
+  };
+
+  const search = (page, searchText): void => {
+    setIsLoading(true);
+    setTableState({
+      ...tableState,
+      count: counts.searchCount,
+    });
+    dispatch(indexUsers({ page, role, active, search: searchText })).then(
```

Comment:
@ekowidianto is `@users.to_a.count` performant? Or would an additional query like [here](https://stackoverflow.com/a/21031792) be better? Because we need to conditionally count the hash if there's a `GROUP BY` clause, but a `.count` without `GROUP BY` just returns an integer.

Reply:
I think we can just conditionally do the count. ‘to_a’ doesnt solve the problem as you are also loading the data into rails memory.